Splitter extension component

stsrki opened the pull request:
Extension from @smfields 

Closes #4286

stsrki commented:
@David-Moreira friendly reminder

David-Moreira commented:
> @David-Moreira friendly reminder

Let me see if I can go over this on the weekend since gotta take it for a spin.

David-Moreira commented:
Both splitters in the demo have different behaviours for how they handle the content inside each section.

**Start section gets squished:**
![image](https://github.com/Megabit/Blazorise/assets/22283161/230c3d2a-50ed-4acf-a5d9-5e14b830e2ce)

**End Section keeps some width relative to existing content:**
![image](https://github.com/Megabit/Blazorise/assets/22283161/da00ea3b-225e-48bb-9f49-b3b8421cd1ce)

David-Moreira commented:
Unrelated to this PR, we can open an issue. Any reason why this isn't fully alphabetical?
![image](https://github.com/Megabit/Blazorise/assets/22283161/fb2e63d8-d5c5-4947-b6b1-dc5b9c8d0048)

David-Moreira commented:
The examples are very simple, and don't make use of parameters, there's a few new apis like:
![image](https://github.com/Megabit/Blazorise/assets/22283161/2317d290-66c3-47c6-bc81-052105840eed)
which leave me wondering or they are supposed to be used and why do we need this extra complexity over regular value types. And does seem not very Blazorise like. I'm guessing this is supposed to be a way to set multiple sections sizes? In Blazorise we prefer to define on each child component which then should inform the parent component of their sizes?

David-Moreira commented on Source/Extensions/Blazorise.Splitter/wwwroot/blazorise.splitter.js:
```diff
@@ -0,0 +1,81 @@
+ * @param {number|number[]|string|undefined} value Option that is a number or array of numbers
+ * @return {number|number[]|undefined} Parsed option
+ */
+function parseNumberOrArray(value) {
```
Lot of stuff here that almost feels like it should be shared stuff? Not sure.

David-Moreira commented on Source/Extensions/Blazorise.Splitter/wwwroot/blazorise.splitter.js:
```diff
@@ -0,0 +1,81 @@
+ * @return A split instance
+ */
+export function initializeSplit(elements, options) {
+    return Split(elements, parseOptions(options));
```
We usually call init with an elementid, keep the instance in memory to get rid of it later, etc....
Does this handle multiple split instances without conflicts?

stsrki replied:
Yes, it works pretty well.

David-Moreira commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,184 @@
+    /// <summary>
+    /// Minimum size of each element.
+    /// </summary>
+    [Parameter] public JavascriptNumberOrArray MinSize { get; init; }
```
I would prefer users to have simple apis they are used to, then these newer ones, because it seems to me we might not need them.
If we really need them, it would be great if they could only be used internally. That is, if we do need them.

David-Moreira commented:
Definitely would like to see more examples.
Something like split.js has for our demo page to test every attribute, just like we do in a few other extension components.
![image](https://github.com/Megabit/Blazorise/assets/22283161/b798d0d4-d19f-4037-a1fc-e2bec771bfed)

And a few standard on our docs, with sizes defined, with multiple splits being used both horizontal and vertical at the same time (if supported), etc...

smfields commented:
> Definitely would like to see more examples. Something like split.js has for our demo page to test every attribute, just like we do in a few other extension components. ![image](https://user-images.githubusercontent.com/22283161/255323965-b798d0d4-d19f-4037-a1fc-e2bec771bfed.png)
> 
> And a few standard on our docs, with sizes defined, with multiple splits being used both horizontal and vertical at the same time (if supported), etc...

Hey David, just wanted to call out that the wrapper has it's own clone of the page you mentioned that you guys might be able to copy for the demo page.

https://blazor-split.smfields.net/[https://blazor-split.smfields.net/](https://blazor-split.smfields.net/)

smfields commented:
> The examples are very simple, and don't make use of parameters, there's a few new apis like: ![image](https://user-images.githubusercontent.com/22283161/255322748-2317d290-66c3-47c6-bc81-052105840eed.png) which leave me wondering or they are supposed to be used and why do we need this extra complexity over regular value types. And does seem not very Blazorise like. I'm guessing this is supposed to be a way to set multiple sections sizes? In Blazorise we prefer to define on each child component which then should inform the parent component of their sizes?

Trying to remember way back here, but I think the main reason I implemented these in the library was so that it handled infinity correctly. 

Infinity has a different behavior to just setting a really big number. Can't remember for sure though.

A lot of what I did on the wrapper was to try to keep things as similar as possible to the JS version. Obviously your goals with this extension might be different.

David-Moreira commented:
> > The examples are very simple, and don't make use of parameters, there's a few new apis like: ![image](https://user-images.githubusercontent.com/22283161/255322748-2317d290-66c3-47c6-bc81-052105840eed.png) which leave me wondering or they are supposed to be used and why do we need this extra complexity over regular value types. And does seem not very Blazorise like. I'm guessing this is supposed to be a way to set multiple sections sizes? In Blazorise we prefer to define on each child component which then should inform the parent component of their sizes?
> 
> Trying to remember way back here, but I think the main reason I implemented these in the library was so that it handled infinity correctly.
> 
> Infinity has a different behavior to just setting a really big number. Can't remember for sure though.
> 
> A lot of what I did on the wrapper was to try to keep things as similar as possible to the JS version. Obviously your goals with this extension might be different.

Yea we'd like to try and make it simple as possible for users. If it can be represented with value types with no limitations that would be great. 
I believe there's supposed to be constants that represent infinity, etc. No idea if you would have some limitation but in theory you could still just check the value type of infinity and translate to whatever you need internally? 
![image](https://github.com/Megabit/Blazorise/assets/22283161/db9e6978-353b-44fa-9870-a3038f86270c)

As for having multiple defined parameters for several sections, I think it's best to be defined in each of the individual sections. At least that's how we've been doing it in similar components and it would be more similar to what users are used to using our library.

Thanks for giving your input on these matters!

stsrki commented on Source/Extensions/Blazorise.Splitter/JsonConverters/CamelCaseEnumJsonConverter.cs:
```diff
@@ -0,0 +1,15 @@
+/// <summary>
+/// JsonConverter for enums that converts to and from a Camel Case result
+/// </summary>
+public class CamelCaseEnumJsonConverter : JsonStringEnumConverter
+{
+    /// <inheritdoc />
+    public CamelCaseEnumJsonConverter() : base( JsonNamingPolicy.CamelCase )
+    {
+    }
+}
```
Users could have their own rules for serializing enums. Meaning it would mess with our enum names. This way, we are strictly defining how it should be serialized. We definitely need this in other places in Blazorise.

David-Moreira replied:
For a moment I forgot you can't define serializer options for specific JSInterop calls. Or at least as far as I know. yea, It's fine then.

stsrki commented on Source/Extensions/Blazorise.Splitter/wwwroot/blazorise.splitter.js:
```diff
@@ -0,0 +1,81 @@
+import './vendors/split.js?v=1.1.3.0';
+
+document.getElementsByTagName("head")[0].insertAdjacentHTML("beforeend", "<link rel=\"stylesheet\" href=\"_content/Blazorise.Splitter/blazorise.splitter.css?v=1.1.3.0\" />");
```
We do, actually. For example, on Animate, Cropper, Markown, Video, etc.

David-Moreira replied:
I see, my bad then, I spoke from memory. I was thinking if users have static files being served from elsewhere, this will probably fail. Anyway they can still load it manually.

stsrki commented:
The `JavascriptNumber` might be removed. We need to see if it makes sense to have it instead of a primitive type. But, `JavascriptNumberOrArray` can stay. JS libraries often expect a single valid, or an array, which `JavascriptNumberOrArray` seems to solve for us.

On the other hand, we have a similar feature on Charts named `IndexableOption`. It does the same thing. Hold a single value or an array, and serialize it based on what it holds.

stsrki commented:
PS. I think I like `JavascriptNumber` the more I think about it. It solves several potential problems for us in the future. And besides, for the user, it should be seamlessly used since it has implicit operators for basic numeric types.

stsrki commented on Source/Extensions/Blazorise.Splitter/JSSplitModule.cs:
```diff
@@ -0,0 +1,46 @@
+    /// <param name="sections">Sections of the splitter</param>
+    /// <param name="options">Splitter configuration options</param>
+    /// <returns>A <see cref="IJSObjectReference"/> to the Animation object</returns>
+    public virtual ValueTask<IJSObjectReference> InitializeSplit( IEnumerable<ElementReference> sections, SplitterOptions options )
```
I tried to cache instances for each splitter in JS, but overall it becomes harder to manage. Also, there is no way to update split instances, it must be created every time something changes. So we can leave it as it is, with the direct reference(IJSObjectReference).

David-Moreira replied:
Maybe update to `InitializeSplitter` then if Initialize can't be used.

stsrki replied:
Done

stsrki commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,184 @@
+
+    protected override async Task OnAfterRenderAsync( bool firstRender )
+    {
+        await CreateInstance();
```
I have optimized it to create only when needed.

stsrki commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,184 @@
+
+                JSSplitInstance = await JSModule.InitializeSplit( splitSections, options );
+            }
+        }
```
There is no update like in other libraries. It has to be recreated every time. This is OK in this situation.

stsrki commented:
OK, I have removed all direct-to-user APIs that mention `JavascriptNumber`. Some APIs like MinSize and MaxSIze are now moved to `SplitterSection` instead of having them on the `Splitter` component. This way user will be able to directly configure each section. Also, every parameter is now based on primitive types, eg. `double`.

Internally I still use `JavascriptNumber` via the `SplitterOptions`. This is to ensure that values like `Infinity` will be properly serialized.

@smfields don't want to bother you too much. It would be great if you could review the new API and let us your thoughts. Thanks

David-Moreira commented:
@stsrki I still see a lot of unresolved comments, do you still want me to re-review?

stsrki commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,184 @@
+
+        try
+        {
+            if ( JSSplitInstance is not null )
+                await JSSplitInstance.InvokeVoidAsync( "destroy" );
```
I have already explained this in other comments. Skipping this.

stsrki commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,184 @@
+                    DragInterval = DragInterval,
+                    Direction = Direction,
+                    Cursor = Cursor,
+                };
```
There were problems with chart options when the serializer was used to serialize nullable fields. In this case, it is not a problem.

stsrki commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,184 @@
+/// <summary>
+/// The Split component allows splitting an area into multiple resizable sections.
+/// </summary>
+public partial class Splitter : BaseComponent, IAsyncDisposable
```
It is OK. We still want for users to be able to define stuff like margin, padding, shadow, etc. It's just a simple `div` for the actual splitter container.

stsrki commented on Source/Extensions/Blazorise.Splitter/SplitterSection.razor.cs:
```diff
@@ -0,0 +1,68 @@
+    /// <inheritdoc />
+    protected override void BuildClasses( ClassBuilder builder )
+    {
+        builder.Append( "split-section" );
```
No need for prefixes.

stsrki replied:
Generally, I try to stick to the naming of the library that we use behind the scenes. In this case, there is no prefix because the split library doesn't have it. Similarly is the case for Bootstrap, Bulma, antdesign. They all should have naming conventions for each of them separately.

The `b-` prefix should be used only in `Blazorise.csproj`, a core library. This is because in there we don't have any knowledge of which CSS framework is running. So we add a prefix to make it more explicit.

David-Moreira replied:
Gotcha going to update.

> The b- prefix should be used only in Blazorise.csproj, a core library. This is because in there we don't have any knowledge of which CSS framework is running. So we add a prefix to make it more explicit.

Uhm, I'm not sure you explained this correctly, with the class builder abstraction you can make it so whatever css framework is running will apply correctly, in fact, that's precisely what you do independent of the library. But I understand what you mean.

David-Moreira replied:
Also we've already used b- outside of the core library, including, for example cropper js which I believe uses a third party library:
![image](https://github.com/Megabit/Blazorise/assets/22283161/8e2bb27c-5a9a-4e1a-96cc-b0e09c3793ce)
So I'm not sure we are exactly following whatever convention you intended us to follow.
And even with the explanation you just gave, I'm still not 100% clear.

David-Moreira replied:
Anyway, updated with this for now:
![image](https://github.com/Megabit/Blazorise/assets/22283161/bb1d25c5-d9b5-40d0-9482-7db9141bad26)

https://github.com/Megabit/Blazorise/wiki/Contributing

David-Moreira replied:
@stsrki Just a reminder that the classes I was advocating to prefix **b-** I believe aren't from the 3rd party library, but "our own"
![image](https://github.com/Megabit/Blazorise/assets/22283161/7358c05a-fb91-4ca8-8109-4d83806ec259)

stsrki replied:
True, but there is only one splitter component. If there were multiple then the prefix would make sense.

stsrki replied:
Haha, I know it's confusing. I should have stuck to one naming convention in the past. We can discuss this in our next meeting for sure.

As I said, `b-` prefix should **generally** be used for stuff where we don't have other means to distinguish a CSS selector or need to add more context to the component. For example, DataGrid does this because it is shared between different frameworks. Autocomplete also. They both use the core Blazorise components.

Cropper is a standalone library. I chose prefixes on the fly basically. Maybe it is a good maybe not so good choice 💁‍♂️

David-Moreira replied:
Yep still don't understand the convention. 
Either it's just me not understanding, or the convention isn't easy to follow.

As a guideline for people contributing to follow, I'm not sure they'll get it right.
But as long as you're reviewing PRs it's ok, I guess... haha

stsrki replied:
That's what I'm telling the entire time, haha. There is only _general_ convention. But mostly, I go with the flow.

For really naming convention, we will both have to sit down and discuss it. It is beyond this PR.

stsrki commented:
> @stsrki I still see a lot of unresolved comments, do you still want me to re-review?

I missed them, sorry. All are now resolved except for customizing cursor and gutter image. I need to see how to do that. You can proceed with review.

stsrki commented on Source/Extensions/Blazorise.Splitter/wwwroot/blazorise.splitter.css:
```diff
@@ -0,0 +1,29 @@
+}
+
+    .gutter.gutter-horizontal {
+        background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAeCAYAAADkftS9AAAAIklEQVQoU2M4c+bMfxAGAgYYmwGrIIiDjrELjpo5aiZeMwF+yNnOs5KSvgAAAABJRU5ErkJggg==');
```
Added GutterBackgroundImage parameter.

stsrki commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,184 @@
+    /// <summary>
+    /// Cursor to display while dragging.
+    /// </summary>
+    [Parameter] public string Cursor { get; init; }
```
It's hard to do with an enum because the whole idea is that the user can practically define any cursor name. And I don't want to have an enum for all of them. I think it would be better to leave it like this.

stsrki commented:
@David-Moreira All comments should now be resolved.

David-Moreira commented:
These range indicators give me no idea of what I am selecting, it's some arbitary number.
In my opninion there should be either
- an indication of the current value **or**
- minimum / max value
This way the user at least can tell the values.
![image](https://github.com/Megabit/Blazorise/assets/22283161/34c696bc-6d21-4f34-a542-b297ad0e9910)

David-Moreira commented:
This thing still happens at minimum, but maybe it's okay? It's expected the user defines whatever appropriate min?

Start side gets squished
![image](https://github.com/Megabit/Blazorise/assets/22283161/59a437c7-e80d-4d85-b464-473644c8d267)
vs
End side does not
![image](https://github.com/Megabit/Blazorise/assets/22283161/6362baef-4d86-4675-b8e3-6d3a1ba28492)

David-Moreira commented:
Is the user expected to define these? Is it something that makes it so the Splitter might not work correctly? Should this just be handled internally?
![image](https://github.com/Megabit/Blazorise/assets/22283161/8a1802bc-06f5-43b3-b656-d51e63821f92)

David-Moreira commented on Source/Extensions/Blazorise.Splitter/JSSplitModule.cs:
```diff
@@ -0,0 +1,47 @@
+    /// <param name="sectionElementRefs">Sections of the splitter.</param>
+    /// <param name="splitterOptions">Splitter configuration options.</param>
+    /// <param name="splitterGutterOptions">Gutter configuration options</param>
+    /// <returns>A <see cref="IJSObjectReference"/> to the Animation object.</returns>
```
copy paste typo maybe. "to the Animation object."

David-Moreira commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,255 @@
+namespace Blazorise.Splitter;
+
+/// <summary>
+/// The Split component allows splitting an area into multiple resizable sections.
```
**Split  => Splitter**, maybe run a Find&Replace

David-Moreira commented on Source/Extensions/Blazorise.Splitter/SplitterSection.razor.cs:
```diff
@@ -0,0 +1,109 @@
+    protected override void OnInitialized()
+    {
+        if ( Parent is null )
+            throw new ArgumentNullException( nameof( Parent ), nameof( SplitterSection ) + " must exist within a Split" );
```
"must exist within a Split"; you mean must exist within Splitter? use **nameof**

David-Moreira commented on Source/Extensions/Blazorise.Splitter/Splitter.razor.cs:
```diff
@@ -0,0 +1,255 @@
+    /// <summary>
+    /// Defines the custom background image for the gutter element.
+    /// </summary>
+    [Parameter] public string GutterBackgroundImage { get; init; }
```
Can we have an example of this on the docs? I think it might be a common operation.

David-Moreira replied:
Additionally tell the user what's expected here, a base64? a path?

David-Moreira reviewed (changes requested):
Working really great! Nice job @smfields , @stsrki 

Also tested dynamically adding / removing sections, no apparent issues.

I really do prefer the value types as the surface api, it seems like it's easier to understand for a regular user then figuring out the "javascript" types (even if they do have implicit operators, I don't think it would be as clear)

David-Moreira commented:
@stsrki Can you test an example similar to this?
                ```
<Splitter Direction="@direction" GutterSize="@gutterSize" Width="@(direction == SplitterDirection.Horizontal ? Width.Is100 : null)" Height="@(direction == SplitterDirection.Vertical ? Height.Is100 : null)">
                    <SplitterSection Index=1 MinSize="@minSize">
                        <div>My Index is 1</div>
                    </SplitterSection>
                    <SplitterSection Index=6 Size="10">
                        <div>My Index is 6</div>
                    </SplitterSection>
                    @foreach ( var item in sections )
                    {
                        <SplitterSection Index=3 MinSize="@minSize">
                            <div>@item</div>
                            <Div>Index is 3</Div>
                        </SplitterSection>
                    }
                    <SplitterSection Index=5 MinSize="@minSize">
                        <div>My Index is 5</div>
                    </SplitterSection>
                </Splitter>
```

I don't  think the indexes are working as expected, the idea should be to reorder the splitters visually right? Otherwise there's no point in it existing? If there is, then it should clearly be explained. Also setting Size specifically in a section, seems like it throws all the other sections off? See prints below:

Defined indexes but apparently not respected:
![image](https://github.com/Megabit/Blazorise/assets/22283161/2089854f-cf3f-4830-9dc1-ac61e7ceb428)


Defined Size of 10 in "Index6" section
![image](https://github.com/Megabit/Blazorise/assets/22283161/39920123-4b77-4a22-8569-071d827931cb)

stsrki commented:
> Is the user expected to define these? Is it something that makes it so the Splitter might not work correctly? Should this just be handled internally? ![image](https://user-images.githubusercontent.com/22283161/256168111-8a1802bc-06f5-43b3-b656-d51e63821f92.png)

This is mainly for our configurable demo to work properly. But generally, users should define a size for the splitter container. Especially for vertical mode because height must be known for it to work properly.

stsrki commented:
> This thing still happens at minimum, but maybe it's okay? It's expected the user defines whatever appropriate min?
> 
> Start side gets squished ![image](https://user-images.githubusercontent.com/22283161/256166855-59a437c7-e80d-4d85-b464-473644c8d267.png) vs End side does not ![image](https://user-images.githubusercontent.com/22283161/256166940-6362baef-4d86-4675-b8e3-6d3a1ba28492.png)

The split works like that. There are no constraint by default, see https://github.com/nathancahill/split/tree/master/packages/splitjs#important-note

David-Moreira commented:
> > This thing still happens at minimum, but maybe it's okay? It's expected the user defines whatever appropriate min?
> > Start side gets squished ![image](https://user-images.githubusercontent.com/22283161/256166855-59a437c7-e80d-4d85-b464-473644c8d267.png) vs End side does not ![image](https://user-images.githubusercontent.com/22283161/256166940-6362baef-4d86-4675-b8e3-6d3a1ba28492.png)
> 
> The split works like that. There are no constraint by default, see https://github.com/nathancahill/split/tree/master/packages/splitjs#important-note

Well then the end should also get squished in my opnion. Behaviour should be the same. Anyway I'm fine levaing as is, it's not a problem I would say.

David-Moreira commented:
> > Is the user expected to define these? Is it something that makes it so the Splitter might not work correctly? Should this just be handled internally? ![image](https://user-images.githubusercontent.com/22283161/256168111-8a1802bc-06f5-43b3-b656-d51e63821f92.png)
> 
> This is mainly for our configurable demo to work properly. But generally, users should define a size for the splitter container. Especially for vertical mode because height must be known for it to work properly.

Do we need a note on the docs  then?

stsrki commented:
> @stsrki Can you test an example similar to this?
> 
> ```
> <Splitter Direction="@direction" GutterSize="@gutterSize" Width="@(direction == SplitterDirection.Horizontal ? Width.Is100 : null)" Height="@(direction == SplitterDirection.Vertical ? Height.Is100 : null)">
>                    <SplitterSection Index=1 MinSize="@minSize">
>                        <div>My Index is 1</div>
>                    </SplitterSection>
>                    <SplitterSection Index=6 Size="10">
>                        <div>My Index is 6</div>
>                    </SplitterSection>
>                    @foreach ( var item in sections )
>                    {
>                        <SplitterSection Index=3 MinSize="@minSize">
>                            <div>@item</div>
>                            <Div>Index is 3</Div>
>                        </SplitterSection>
>                    }
>                    <SplitterSection Index=5 MinSize="@minSize">
>                        <div>My Index is 5</div>
>                    </SplitterSection>
>                </Splitter>
> ```
> 
> I don't think the indexes are working as expected, the idea should be to reorder the splitters visually right? Otherwise there's no point in it existing? If there is, then it should clearly be explained. Also setting Size specifically in a section, seems like it throws all the other sections off? See prints below:
> 
> Defined indexes but apparently not respected: ![image](https://user-images.githubusercontent.com/22283161/256179666-2089854f-cf3f-4830-9dc1-ac61e7ceb428.png)
> 
> Defined Size of 10 in "Index6" section ![image](https://user-images.githubusercontent.com/22283161/256179410-39920123-4b77-4a22-8569-071d827931cb.png)

My idea with Index was to make sure that SplitterSection settings are properly indexed and sent to create a Split instance. It didn't occur to me to use it for the rendering order.

David-Moreira commented:
> > @stsrki Can you test an example similar to this?
> > ```
> > <Splitter Direction="@direction" GutterSize="@gutterSize" Width="@(direction == SplitterDirection.Horizontal ? Width.Is100 : null)" Height="@(direction == SplitterDirection.Vertical ? Height.Is100 : null)">
> >                    <SplitterSection Index=1 MinSize="@minSize">
> >                        <div>My Index is 1</div>
> >                    </SplitterSection>
> >                    <SplitterSection Index=6 Size="10">
> >                        <div>My Index is 6</div>
> >                    </SplitterSection>
> >                    @foreach ( var item in sections )
> >                    {
> >                        <SplitterSection Index=3 MinSize="@minSize">
> >                            <div>@item</div>
> >                            <Div>Index is 3</Div>
> >                        </SplitterSection>
> >                    }
> >                    <SplitterSection Index=5 MinSize="@minSize">
> >                        <div>My Index is 5</div>
> >                    </SplitterSection>
> >                </Splitter>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't think the indexes are working as expected, the idea should be to reorder the splitters visually right? Otherwise there's no point in it existing? If there is, then it should clearly be explained. Also setting Size specifically in a section, seems like it throws all the other sections off? See prints below:
> > Defined indexes but apparently not respected: ![image](https://user-images.githubusercontent.com/22283161/256179666-2089854f-cf3f-4830-9dc1-ac61e7ceb428.png)
> > Defined Size of 10 in "Index6" section ![image](https://user-images.githubusercontent.com/22283161/256179410-39920123-4b77-4a22-8569-071d827931cb.png)
> 
> My idea with Index was to make sure that SplitterSection settings are properly indexed and sent to create a Split instance. It didn't occur to me to use it for the rendering order.

Right, but if it's there, as a user I would assume it would reorder, if we have no intentions of that, then we should explicitly mention it.

Also I argue that if it's only for re-ordering the settings, then it's redudant to have it?

David-Moreira commented:
@stsrki Is this working as intended?
![image](https://github.com/Megabit/Blazorise/assets/22283161/b8a177f8-afee-4b92-96b1-19722bad98fb)

David-Moreira commented:
This
```
<Splitter Direction="@direction" GutterSize="@gutterSize" Width="@(direction == SplitterDirection.Horizontal ? Width.Is100 : null)" Height="@(direction == SplitterDirection.Vertical ? Height.Is100 : null)">
                    <SplitterSection MinSize="@minSize">
                        <div>My Index is 1</div>
                    </SplitterSection>
                    <SplitterSection Size="10">
                        <div>My Index is 6</div>
                    </SplitterSection>
                    <SplitterSection  MinSize="@minSize">
                        <div>My Index is 5</div>
                    </SplitterSection>
                </Splitter>
```
Results in this:
![image](https://github.com/Megabit/Blazorise/assets/22283161/72ff93e1-d4ee-43de-a360-bcd6c279b7e7)

Something seems off to me. I don't think the middle section is respecting the Size 10? Or since it does not have a minsize defined, it's just whatever? Up to the underlying lib to manage this?

David-Moreira commented:
While I was writing my last comment, I noticed this:
![image](https://github.com/Megabit/Blazorise/assets/22283161/44c92c47-c649-443b-a94f-88b13651e18d)
Stack trace:
```
      Unhandled exception rendering component: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
      Microsoft.JSInterop.JSDisconnectedException: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
         at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(Int64 asyncHandle, String identifier, String argsJson, JSCallResultType resultType, Int64 targetInstanceId)
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, CancellationToken cancellationToken, Object[] args)
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
         at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
         at Microsoft.JSInterop.Implementation.JSObjectReference.DisposeAsync()
         at Blazorise.Splitter.Splitter.DisposeAsync(Boolean disposing) in C:\Users\david\source\repos\Blazorise\Source\Extensions\Blazorise.Splitter\Splitter.razor.cs:line 100
         at Blazorise.BaseAfterRenderComponent.DisposeAsync() in C:\Users\david\source\repos\Blazorise\Source\Blazorise\Base\BaseAfterRenderComponent.cs:line 133
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass74_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
fail: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
      Unhandled exception in circuit 'c7k09NHyBwLwXfkEW2Jo9gw-nW6k_i4MZXCdes-HBIw'.
      Microsoft.JSInterop.JSDisconnectedException: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
         at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(Int64 asyncHandle, String identifier, String argsJson, JSCallResultType resultType, Int64 targetInstanceId)
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, CancellationToken cancellationToken, Object[] args)
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
         at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
         at Microsoft.JSInterop.Implementation.JSObjectReference.DisposeAsync()
         at Blazorise.Splitter.Splitter.DisposeAsync(Boolean disposing) in C:\Users\david\source\repos\Blazorise\Source\Extensions\Blazorise.Splitter\Splitter.razor.cs:line 100
         at Blazorise.BaseAfterRenderComponent.DisposeAsync() in C:\Users\david\source\repos\Blazorise\Source\Blazorise\Base\BaseAfterRenderComponent.cs:line 133
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass74_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
```

stsrki commented:
> @stsrki Is this working as intended? ![image](https://user-images.githubusercontent.com/22283161/256253800-b8a177f8-afee-4b92-96b1-19722bad98fb.png)

Yes. If only one section has a default size then the last section will be `infinite`. If there is any section in between they will resize automatically to their minimum. This is how split.js works.

stsrki commented:
> This
> 
> ```
> <Splitter Direction="@direction" GutterSize="@gutterSize" Width="@(direction == SplitterDirection.Horizontal ? Width.Is100 : null)" Height="@(direction == SplitterDirection.Vertical ? Height.Is100 : null)">
>                     <SplitterSection MinSize="@minSize">
>                         <div>My Index is 1</div>
>                     </SplitterSection>
>                     <SplitterSection Size="10">
>                         <div>My Index is 6</div>
>                     </SplitterSection>
>                     <SplitterSection  MinSize="@minSize">
>                         <div>My Index is 5</div>
>                     </SplitterSection>
>                 </Splitter>
> ```
> 
> Results in this: ![image](https://user-images.githubusercontent.com/22283161/256254092-72ff93e1-d4ee-43de-a360-bcd6c279b7e7.png)
> 
> Something seems off to me. I don't think the middle section is respecting the Size 10? Or since it does not have a minsize defined, it's just whatever? Up to the underlying lib to manage this?

The `Size` is only the default size. After that it can be resized to anything unless it is constraint by the Min or Max sizes.

stsrki commented:
> While I was writing my last comment, I noticed this: ![image](https://user-images.githubusercontent.com/22283161/256254691-44c92c47-c649-443b-a94f-88b13651e18d.png) Stack trace:
> 
> ```
>       Unhandled exception rendering component: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
>       Microsoft.JSInterop.JSDisconnectedException: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
>          at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(Int64 asyncHandle, String identifier, String argsJson, JSCallResultType resultType, Int64 targetInstanceId)
>          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, CancellationToken cancellationToken, Object[] args)
>          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
>          at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
>          at Microsoft.JSInterop.Implementation.JSObjectReference.DisposeAsync()
>          at Blazorise.Splitter.Splitter.DisposeAsync(Boolean disposing) in C:\Users\david\source\repos\Blazorise\Source\Extensions\Blazorise.Splitter\Splitter.razor.cs:line 100
>          at Blazorise.BaseAfterRenderComponent.DisposeAsync() in C:\Users\david\source\repos\Blazorise\Source\Blazorise\Base\BaseAfterRenderComponent.cs:line 133
>          at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass74_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
> fail: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
>       Unhandled exception in circuit 'c7k09NHyBwLwXfkEW2Jo9gw-nW6k_i4MZXCdes-HBIw'.
>       Microsoft.JSInterop.JSDisconnectedException: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
>          at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(Int64 asyncHandle, String identifier, String argsJson, JSCallResultType resultType, Int64 targetInstanceId)
>          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, CancellationToken cancellationToken, Object[] args)
>          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
>          at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
>          at Microsoft.JSInterop.Implementation.JSObjectReference.DisposeAsync()
>          at Blazorise.Splitter.Splitter.DisposeAsync(Boolean disposing) in C:\Users\david\source\repos\Blazorise\Source\Extensions\Blazorise.Splitter\Splitter.razor.cs:line 100
>          at Blazorise.BaseAfterRenderComponent.DisposeAsync() in C:\Users\david\source\repos\Blazorise\Source\Blazorise\Base\BaseAfterRenderComponent.cs:line 133
>          at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass74_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
> ```

Can you replicate?

David-Moreira commented:
> > While I was writing my last comment, I noticed this: ![image](https://user-images.githubusercontent.com/22283161/256254691-44c92c47-c649-443b-a94f-88b13651e18d.png) Stack trace:
> > ```
> >       Unhandled exception rendering component: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
> >       Microsoft.JSInterop.JSDisconnectedException: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
> >          at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(Int64 asyncHandle, String identifier, String argsJson, JSCallResultType resultType, Int64 targetInstanceId)
> >          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, CancellationToken cancellationToken, Object[] args)
> >          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
> >          at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
> >          at Microsoft.JSInterop.Implementation.JSObjectReference.DisposeAsync()
> >          at Blazorise.Splitter.Splitter.DisposeAsync(Boolean disposing) in C:\Users\david\source\repos\Blazorise\Source\Extensions\Blazorise.Splitter\Splitter.razor.cs:line 100
> >          at Blazorise.BaseAfterRenderComponent.DisposeAsync() in C:\Users\david\source\repos\Blazorise\Source\Blazorise\Base\BaseAfterRenderComponent.cs:line 133
> >          at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass74_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
> > fail: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
> >       Unhandled exception in circuit 'c7k09NHyBwLwXfkEW2Jo9gw-nW6k_i4MZXCdes-HBIw'.
> >       Microsoft.JSInterop.JSDisconnectedException: JavaScript interop calls cannot be issued at this time. This is because the circuit has disconnected and is being disposed.
> >          at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(Int64 asyncHandle, String identifier, String argsJson, JSCallResultType resultType, Int64 targetInstanceId)
> >          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, CancellationToken cancellationToken, Object[] args)
> >          at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
> >          at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
> >          at Microsoft.JSInterop.Implementation.JSObjectReference.DisposeAsync()
> >          at Blazorise.Splitter.Splitter.DisposeAsync(Boolean disposing) in C:\Users\david\source\repos\Blazorise\Source\Extensions\Blazorise.Splitter\Splitter.razor.cs:line 100
> >          at Blazorise.BaseAfterRenderComponent.DisposeAsync() in C:\Users\david\source\repos\Blazorise\Source\Blazorise\Base\BaseAfterRenderComponent.cs:line 133
> >          at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass74_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
> > ```
> 
> Can you replicate?

Just having the server running after closing the circuit seems to eventually cause it.
